nemo/core/config/pytorch_lightning.py

```diff
@@ -13,39 +13,82 @@
 # limitations under the License.
 
 from dataclasses import dataclass
-from typing import Any, Dict, List, Optional
+from typing import Any, Dict, List, Optional, Union
+
+from hydra.core.config_store import ConfigStore
+from hydra.types import ObjectConf
 
 __all__ = ['TrainerConfig']
 
 
+cs = ConfigStore.instance()
+
+
 @dataclass
 class TrainerConfig:
     """
     Configuration of PyTorch Lightning Trainer.
-
     It is not derived from Config as it is not a NeMo object (and in particular it doesn't need a name).
-
     ..warning:
         Picked just few params of the PTL trainer for now. This needs to be discussed.
-
     ..note:
         For the details on the function/meanings of the arguments, please refer to:
         https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#
     """
 
+    logger: Any = True
+    checkpoint_callback: Any = True
+    early_stop_callback: Any = False
+    callbacks: Optional[Any] = None
+    default_root_dir: Optional[str] = None
     gradient_clip_val: float = 0
     process_position: int = 0
     num_nodes: int = 1
     num_processes: int = 1
-    gpus: Optional[int] = None
+    gpus: Optional[Any] = None
     auto_select_gpus: bool = False
+    tpu_cores: Optional[Any] = None
     log_gpu_memory: Optional[str] = None
     progress_bar_refresh_rate: int = 1
+    overfit_batches: Any = 0.0
+    track_grad_norm: Any = -1
     check_val_every_n_epoch: int = 1
     fast_dev_run: bool = False
+    accumulate_grad_batches: Any = 1
     max_epochs: int = 1000
     min_epochs: int = 1
-    distributed_backend: Optional[str] = None
     max_steps: Optional[int] = None
-    accumulate_grad_batches: int = 1
-    amp_level: str = "O0"
+    min_steps: Optional[int] = None
+    limit_train_batches: Any = 1.0
+    limit_val_batches: Any = 1.0
+    limit_test_batches: Any = 1.0
+    val_check_interval: Any = 1.0
+    log_save_interval: int = 100
+    row_log_interval: int = 50
+    distributed_backend: Optional[str] = None
+    sync_batchnorm: bool = False
+    precision: int = 32
+    weights_summary: Optional[str] = "full"  # ModelSummary.MODE_DEFAULT
+    weights_save_path: Optional[str] = None
+    num_sanity_val_steps: int = 2
+    truncated_bptt_steps: Optional[int] = None
+    resume_from_checkpoint: Optional[str] = None
+    profiler: Optional[Any] = None
+    benchmark: bool = False
+    deterministic: bool = False
+    reload_dataloaders_every_epoch: bool = False
+    auto_lr_find: Any = False
+    replace_sampler_ddp: bool = True
+    terminate_on_nan: bool = False
+    auto_scale_batch_size: Any = False
+    prepare_data_per_node: bool = True
+    amp_backend: str = 'native'
+    amp_level: str = 'O2'  # backward compatible, todo: remove in v1.0.0
+
+
+# Register the trainer config.
+cs.store(
+    group="trainer",
```

Comment:
Why is this registration necessary? Why not perform the config merge similar to what @blisc does inside Experiment Manager?

Reply:
This is standard hydra mechanism for registering configs that can be used them using e.g. instantiate. This is not important from the point of view of this PR.

Reply:
Thinking about the way of injecting schemas in experiment manager:
https://github.com/NVIDIA/NeMo/blob/b90a75fcbcb792aded754bd96a2082381c3e7d48/nemo/utils/exp_manager.py#L138

Please note at that point you have the configuration object, whereas in hydra_runner you don't... I see limitations of my approach, but not sure how this should work with "validation by manual merging"...

Reply:
I don't think we should utilize hydras approach of config store at all. OmegaConf structured is more useful and general for us. My draft PR next week will attempt to remove this and still support modelPT configs. Let's see where that goes.